bodo/libs/streaming/join.py

```diff
@@ -141,6 +140,38 @@ def key(self):
     def n_keys(self):
         return len(self.build_key_inds)
 
+    def is_precise(self):
+        return (
+            self.build_table_type != types.unknown
+            and self.probe_table_type != types.unknown
+        )
+
+    def unify(self, typingctx, other):
+        """Unify two JoinStateType instances when build_table_type/probe_table_type
+        are not resolved.
+        """
+        if isinstance(other, JoinStateType):
```

Comment:
This is the other unify that would need to persist. However if we add a StreamingStateType `canUnify()` that allows us to just capture the additions to the if check we can simplify the changes per function (including having it default to True).

Reply:
See above.